registryprotocols/ociclient/type_oci_pulled_image.go

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+
+package ociclient
+
+import (
+	"io"
+	"io/fs"
+)
+
+// PulledOCIImage is an iterator interface to read pulled OCI images. You will have to call Next()
```

Comment:
Consider copying what `tar.Reader` does? Instead of a bool to signal completion, this could return an `io.EOF` at the end. Also just returning the `tar.Header` is a little nicer than having these methods, but I get why you might not want to expose that implementation detail.

Alternatively, you could use actual iterators in go 1.23: https://tip.golang.org/doc/go1.23#iterators

I played with this here: https://github.com/jonjohnsonjr/itar

Another alternative (probably my preference) would be to expose an `fs.FS` for consumers instead of a homegrown iterator. I've implemented one for myself here: https://github.com/jonjohnsonjr/targz/tree/main/tarfs

Also I've written up some notes about the general approach and its performance https://github.com/jonjohnsonjr/til/blob/main/post/tarfs.md

You would "just" need to implement something that multiplexes across multiple fs.FS implementations (maybe I should do this myself...) to support multiple layers.

(Happy to chat about this more over slack or wherever.)